Fix cascading delete

FengyunPan opened the pull request:
Kubernetes 1.6 adds a PropagationPolicy to the delete options,
obsoleting the previous OrphanDependents boolean. This PR sets
PropagationPolicy=foreground instead of OrphanDependents=false
when federation controller manager delete cluster resource.

@caesarxuchao ptal

**Release note**:
```release-note
NONE
```

madhusudancs commented:
/unassign
/assign @nikhiljindal

caesarxuchao commented:
/assign @nikhiljindal @caesarxuchao

caesarxuchao commented:
Sorry, i'll take a look after release 1.7 ships.

colhom commented on federation/pkg/federatedtypes/crudtester/crudtester.go:
```diff
@@ -122,16 +122,18 @@ func (c *FederatedTypeCRUDTester) CheckUpdate(obj pkgruntime.Object) {
 	}
 
-	deletingInCluster := (orphanDependents != nil && *orphanDependents == false)
+	deletingInCluster := defaultPropagationPolicy != nil && *defaultPropagationPolicy == "Foreground"
```
`.. && *defaultPropagationPolicy == metav1.DeletePropagationForeground`

FengyunPan replied:
Done, thank you.

spiffxp commented:
/ok-to-test

caesarxuchao commented on federation/pkg/federation-controller/deployment/deploymentcontroller.go:
```diff
@@ -186,8 +186,10 @@ func NewDeploymentController(federationClient fedclientset.Interface) *Deploymen
 			rs := obj.(*extensionsv1.Deployment)
-			orphanDependents := false
-			err := client.Extensions().Deployments(rs.Namespace).Delete(rs.Name, &metav1.DeleteOptions{OrphanDependents: &orphanDependents})
+			defaultPropagationPolicy := metav1.DeletePropagationForeground
```
Maybe s/defaultPropagationPolicy/policy? I don't see why it's called "default".

caesarxuchao replied:
@nikhiljindal is the federation control plane going to support both "foreground" and "background" gc? If so, an option should be passed into this function.

FengyunPan replied:
orphanDependents := false
client.Extensions().Deployments(rs.Namespace).Delete(rs.Name, metav1.DeleteOptions{OrphanDependents: &orphanDependents})
Actually, the code can't cascade the deletion of deployment's pods. I will check it later.

caesarxuchao commented:
> Federation controllerManager just removes the federation resource
and orphans the corresponding resources when orphanDependents is
false and PropagationPolicy is nil,

@FengyunPan i don't think this claim is true. The underlying objects are [deleted](https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L131-L155) with orphan=false, so they and their dependents will be garbage collected, in the background.

That said, i think using `DeletePropagationForeground` when deleting the underlying resources is correct, because the federation control plane intends to support foreground GC. @nikhiljindal could you confirm?

FengyunPan commented:
The 'kubelet delete deployment xxx'  work fine, it called "m.RESTClient.Delete().
		NamespaceIfScoped(namespace, m.NamespaceScoped).Resource(m.Resource).Name(name).Body(options).Do().Error()", 
but "c.client.Delete().Namespace(c.ns).Resource("deployments").	VersionedParams(&listOptions, heme.ParameterCodec).	Body(options).	Do().Error()" can't work.
I don't know the relationship between OrphanDependents and PropagationPolicy, is there any doc about it?

caesarxuchao commented:
There should be comment on the types.

The failed cascading deletion was fixed by https://github.com/kubernetes/kubernetes/pull/44058.

FengyunPan commented:
@caesarxuchao Thank you, I have update the comment and PR. 
The fedaration deployment has been merged into 'sync/controller.go'. When user delete fed-deployment, adapter.ClusterDelete() will be called. PTAL.

caesarxuchao commented on test/integration/federation/crud_test.go:
```diff
@@ -49,19 +50,19 @@ func TestFederationCRUD(t *testing.T) {
-	orphanedDependents := true
-	testCases := map[string]*bool{
-		"Resource should not be deleted from underlying clusters when OrphanDependents is true": &orphanedDependents,
-		"Resource should not be deleted from underlying clusters when OrphanDependents is nil":  nil,
```
Why do you remove this case?

FengyunPan replied:
Oops, sorry, I restore the case soon.

caesarxuchao commented:
One nit, otherwise lgtm.

FengyunPan commented:
Set PropagationPolicy default to Background for federation.
When I use Foreground, always timed out to do a deletion.
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/48138/pull-kubernetes-unit/40365/

FengyunPan commented:
/assign @madhusudancs

caesarxuchao commented:
I think Foreground is the right behavior. Could you increase the timeout? Probably the timeout is because it takes time to delete all pods, then the rs, then the deployment.

madhusudancs commented:
@FengyunPan, @nikhiljindal and @caesarxuchao are the best people to review this PR, not me. So I am unassigning myself.

/unassign

caesarxuchao commented:
One outstanding [comment](https://github.com/kubernetes/kubernetes/pull/48138#issuecomment-315145750), otherwise lgtm.

@nikhiljindal it's your turn ;)

nikhiljindal commented on federation/pkg/federatedtypes/crudtester/crudtester.go:
```diff
@@ -139,11 +144,11 @@ func (c *FederatedTypeCRUDTester) CheckDelete(obj pkgruntime.Object, orphanDepen
-		// May need extra time to delete both federation and cluster resources
-		waitTimeout = c.clusterWaitTimeout
-	}
+	waitTimeout := ForeverTimeout
```
Pass the right clusterWaitTimeout in https://github.com/kubernetes/kubernetes/blob/50ec4384ff0ac482beb164767e562765ad82bde7/test/integration/federation/framework/crudtester.go#L46 rather than overwriting it here.

FengyunPan replied:
Oops, I will update it.

nikhiljindal commented on federation/pkg/federatedtypes/crudtester/crudtester.go:
```diff
@@ -139,11 +144,11 @@ func (c *FederatedTypeCRUDTester) CheckDelete(obj pkgruntime.Object, orphanDepen
-	}
+	waitTimeout := ForeverTimeout
+	//if deletingInCluster {
+	//	// May need extra time to delete both federation and cluster resources
```
Delete unused code

FengyunPan replied:
Done

nikhiljindal commented on federation/pkg/federatedtypes/crudtester/crudtester.go:
```diff
@@ -155,7 +160,7 @@ func (c *FederatedTypeCRUDTester) CheckDelete(obj pkgruntime.Object, orphanDepen
 	})
 	if err != nil {
-		c.tl.Fatalf("Error deleting federated %s %q: %v", c.kind, qualifiedName, err)
+		c.tl.Fatalf("Error deleting federated %s %q in %s: %v", c.kind, qualifiedName, waitTimeout, err)
```
%s seconds?

FengyunPan replied:
The waitTimeout contains unit.

nikhiljindal commented on federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go:
```diff
@@ -67,9 +67,11 @@ func NewDeletionHelper(
-// and deletion helper does a cascading deletion.
+// If user deletes the resource with nil DeleteOptions or DeletionOptions.OrphanDependents = true,
+// then the federation controllerManager just delete fed-resource.
+// if user deletes the resource with DeletionOptions.OrphanDependents = true,
```
It should be OrphanDependents = false here instead of true?

FengyunPan replied:
Yes, done.

nikhiljindal commented on federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go:
```diff
@@ -67,9 +67,11 @@ func NewDeletionHelper(
-// DeletionOptions.OrphanDependents = true then the apiserver removes the orphan finalizer
-// and deletion helper does a cascading deletion.
+// If user deletes the resource with nil DeleteOptions or DeletionOptions.OrphanDependents = true,
+// then the federation controllerManager just delete fed-resource.
```
controllerManager -> controller manager.
delete -> deletes
fed-resource -> federation resource

FengyunPan replied:
Thank you for fixing it

nikhiljindal commented on federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go:
```diff
@@ -67,9 +67,11 @@ func NewDeletionHelper(
+// then the federation controllerManager just delete fed-resource.
+// if user deletes the resource with DeletionOptions.OrphanDependents = true,
+// then the federation apiserver removes the orphan finalizer and deletion helper
+// does a cascading deletion(delete fed-resource and delete cluster-resource).
```
does a cascading deletion (deletes federation resource and cluster resources).

FengyunPan replied:
Done.

nikhiljindal commented:
Sorry for the delay in review. I have a few minor nits but at high level, am not sure of the value this adds.

We have 2 options:
* Say that federation only supports OrphanDependents and we do not support PropagationPolicy in federation (there is no code that uses it and there are not tests to ensure that it works). This is true currently.
* Make this change and say that we support Foreground PropagationPolicy, but not Background in federation.

I think the first message is clearer.
If this is just the first step in supporting all PropagationPolicies in federation, then I am fine with this change since then the second message is only temporary.
But without any plan to support Background policy, this change will lead to more confusion, imo.

Happy to hear other opinions.

FengyunPan commented:
@nikhiljindal Deleting federation deployment and cluster deployment(contains rs and pod) is my PR's purpose. Actrually, after deleting federation deployment with OrphanDependents = false, the pods of federation deployment was left behind. Is it need to add Foreground PropagationPolicy into DeleteOptions?

nikhiljindal commented:
aah thanks for explaining that.
Please update the PR description to say that you are fixing that bug. Please update the PR description as well.

I looked at the code again and we are only changing the DELETE request that is sent to clusters, we are now setting PropagationPolicy to Foreground. This is not changing the DELETE request that is sent to federation apiserver by federation controller, hence my comment above is not valid.

This looks good to me. Will add lgtm label, once the open comments have been fixed.
Thanks!

FengyunPan commented:
/retest

FengyunPan commented:
/test pull-kubernetes-unit

FengyunPan commented:
/retest

FengyunPan commented:
@nikhiljindal Hi, can you help me? I just use PropagationPolicy for deleting cluster resource and didn't update the deletion of federation resource. Why the test case times out?

nikhiljindal commented:
Which test case is timing out?

Before this change, deletion of dependent resources in underlying clusters was happening in background. For ex: while deleting federated deployment, federated controller was deleting deployment from underlying cluster such that it does not wait for replicasets and pods to be deleted (they were being deleted later in the background, without blocking deployment deletion). With this change, deployment deletion in underlying cluster will wait for replicasets and pods deletion. So it is expected to take some more time than before.

FengyunPan commented:
@nikhiljindal This case is timing out: k8s.io/kubernetes/test/integration/federation -run TestFederationCRUD

I change more time for this case later.

Commit:
[Federation] Fix cascading delete

After deleting federation deployment with OrphanDependents=false,
the pods of federation deployment was left behind.
Kubernetes 1.6 adds a PropagationPolicy to the delete options,
obsoleting the previous OrphanDependents boolean. This PR sets
PropagationPolicy=foreground instead of OrphanDependents=false
when federation controller manager delete cluster resource.

FengyunPan commented:
/test pull-kubernetes-bazel
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-etcd3

FengyunPan commented:
/retest
